### ISSUE TITLE
<VirtualHost *> problems

### ISSUE THREAD

**Issue by bmw** [certbot/certbot#3981]
Steps to reproduce:

1. Take the default Apache config and change the default vhost from `<VirtualHost *:80>` to `<VirtualHost *>`.
2. Run `certbot -d <domain> --apache` for any domain name pointing at that server.

Domain validation fails. The vhost the Apache plugin makes is something like the following:
```
<IfModule mod_ssl.c>
<VirtualHost *>
    ServerName 996de558c87bf78211dfe71c0cfa7f33.8bfada933e08175738f69ac8321bf840.acme.invalid
    UseCanonicalName on
    SSLStrictSNIVHostCheck on

    LimitRequestBody 1048576

    Include /etc/letsencrypt/options-ssl-apache.conf
    SSLCertificateFile /var/lib/letsencrypt/_Ql9ONFhja57j8csu0ORv8MGshr_1dtCyNa8AuP4h1Q.crt
    SSLCertificateKeyFile /var/lib/letsencrypt/_Ql9ONFhja57j8csu0ORv8MGshr_1dtCyNa8AuP4h1Q.pem

    DocumentRoot /var/lib/letsencrypt/tls_sni_01_page/
</VirtualHost>

</IfModule>
```
Trying to connect to the server using `openssl s_client` for the above `ServerName` causes:
```
140246747047576:error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol
```
At first glance the generated vhost looks right to me. Perhaps the problem is having both HTTP and HTTPS vhosts using `<VirtualHost *>` and we should make the challenge vhost `<VirtualHost *:443>` in this case?

Any thoughts @joohoi?

**Comment by jenlampton** [comment]
Possibly related? https://github.com/certbot/certbot/issues/3103

**Comment by osirisinferi** [comment]
> At first glance the generated vhost looks right to me. Perhaps the problem is having both HTTP and HTTPS vhosts using <VirtualHost *> and we should make the challenge vhost <VirtualHost *:443> in this case?

That is correct. See for more information about the selection process how Apache ends up at which `VirtualHost` section:

* https://httpd.apache.org/docs/2.4/vhosts/name-based.html
* https://httpd.apache.org/docs/2.4/mod/core.html#virtualhost

TL;DR version with my interpretation:

* First, Apache will select the `VirtualHost` corresponding with the most detailed (i.e., non-wildcard) IP and port combination. If such a `VirtualHost` exists, wildcard `VirtualHost`s will be ignored. If more than one `VirtualHost` based on IP/port remains, Apache will look at the `ServerName` directive, but again, ignore the other "wildcard" `VirtualHost`s;
* If such an IP/port match doesn't exist, Apache will look at `VirtualHost` sections with wildcards;
* It *probably* will prefer wildcard IP's with a specified port number above just a wild card without specifying a port number;
* If nothing gives a specific (IP/port) matches, Apache will look at the `ServerName` directive.

As the `tls-sni-01` challenge always uses port 443, I see no harm in always specifying that port?

**Comment by bmw** [comment]
@joohoi are you expecting to get this done for the release next week?

**Comment by joohoi** [comment]
Yes, it should be a short PR to review as well.

**Comment by narcisgarcia** [comment]
I have this same problem in separated scenario (webserver:443 certificateshost:80) sharing single public IP (1.2.3.4) with Apache profiles:
```
<VirtualHost *:443>
   ...
   SSLCertificateFile ".../default.example.net/fullchain.pem
</VirtualHost>
<VirtualHost *:443>
   ServerName      www.example.net
   ...
   SSLCertificateFile ".../www.example.net/fullchain.pem
</VirtualHost>
```
I run this, forwarding only port 80 to renewals host:
`certbot certonly ... --standalone -n -d www.example.net`
But ACME actions seem to include getting any certificate CN found when visiting client IP:
https://1.2.3.4:443/
https://1.2.3.4:80/

1. ACME server should not assume any served certificate and CN when "visiting" client HTTP/S. That certificate could be wrong, obsolete, self-signed, or generic for default one for no-SNI clients.
2. ACME server should forget any :443 information if is switching to another port (:80).

**Comment by narcisgarcia** [comment]
Solution to my particular problem has been:

$ certbot certonly --preferred-challenges http --standalone -n -d www.example.net

**Comment by bmw** [comment]
The issue I described in my original post has been resolved. If you're continuing to have problems here, please open a new issue and fill out the issue template so we can better help you.
